File: exercism/scala/robot-name/src/main/scala/RobotName.scala
import scala.util.Random.nextInt
import scala.util.Random
import collection.mutable

object NameGenerator{
    val alreadySet = mutable.HashSet.empty[String]
    var counter = 0

     def  getName():String = {
        if(counter >= 676000) {
            return "Maxed!"
        }
        val r = new Random(System.currentTimeMillis())
        while(true){            
            val name = f"${('A' to 'Z')(nextInt(26))}${('A' to 'Z')(nextInt(26))}${nextInt(999)}%03d"
            printf("alreadySet=%d, name=%s\n",alreadySet.size, name)
            if(!(alreadySet contains name)){
                counter+=1
                alreadySet += name
                return name                
            }
        }   
        return "Error!"     
    }
}

class Robot{
    var name: String = ""
    reset()
    def reset(){
        name = NameGenerator.getName()
    }

}

File: exercism/kotlin/robot-name/src/main/kotlin/RobotName.kt
// class Robot {
// init {
// reset()
// }
// lateinit var backname: String

// val name: String
// get() = backname
// fun random(isDigit: Boolean): Int {
// if (isDigit) {
// return kotlin.random.Random.nextInt(9) + '0'.toInt()
// } else {
// return kotlin.random.Random.nextInt(26) + 'A'.toInt()
// }
// }
// companion object Set {
// val s = mutableSetOf<String>()
// }
// fun reset() {
// while (true) {
// val temp =
// "AA111"
//// .map {
//// println(it)
//// it
//// }
// .map { random(it == '1').toChar() }
//// .map {
//// println(it)
//// it
//// }
// .joinToString("")
// if (!Robot.s.contains(temp)) {
// backname = temp
// Robot.s.add(temp)
// return
// }
// }
// }

private val random = java.util.Random()

private fun List<*>.randomSample(n: Int) = (0 until n).map { this[random.nextInt(this.size)] }

private val names = mutableSetOf<String>()
private val letters = ('A'..'Z').toList()
private val numbers = ('0'..'9').toList()

class Robot {
    var name: String
    init {
        name = generateName()
    }
    fun reset() {
        name = generateName()
    }
    private tailrec fun generateName(): String {
        val name = letters.randomSample(2).plus(numbers.randomSample(3)).joinToString("")
        return when {
            names.add(name) -> name
            else -> generateName()
        }
    }
}
